extensions/amp-story/1.0/media-performance-metrics-service.js

```diff
@@ -434,4 +418,27 @@ export class MediaPerformanceMetricsService {
     timeStamps.waiting = Date.now();
     mediaEntry.status = Status.WAITING;
   }
+
+  /**
+   * @param {!HTMLMediaElement} media
+   * @return {!CacheState}
+   * @private
+   */
+  getVideoCacheState_(media) {
+    let hasCachedSource = false;
+    const isCachedSource = (source) =>
+      source.hasAttribute('i-amphtml-video-cached-source');
+    // All video caching mechanisms rely on HTMLSourceElements and never a src
+    // on the HTMLMediaElement as it does not allow for fallback sources.
+    const sources = toArray(media.querySelectorAll('source'));
+    for (const source of sources) {
+      if (media.currentSrc === source.src && isCachedSource(source)) {
+        return CacheState.CACHE;
+      }
+      if (isCachedSource(source)) {
+        hasCachedSource = true;
+      }
```

Comment:
Nit: To prevent defining the anon function, you could do
```js
// All video caching mechanisms rely on HTMLSourceElements and never a src
// on the HTMLMediaElement as it does not allow for fallback sources.
const sources = toArray(media.querySelectorAll('source'));

let hasCachedSource = false;
for (const source of sources) {
  if (source.hasAttribute('i-amphtml-video-cached-source')) {
    // You can explain why this returns early so we understand when we look back at it (either this or the method documentation)
    if (media.currentSrc === source.src) {
      return CacheState.CACHE;
    }
    hasCachedSource = true;
  }
}
return hasCachedSource ? CacheState.ORIGIN_CACHE_MISS : CacheState.ORIGIN;
```

Also, keeping the declaration of `hasCachedSource` next to the loop will make it a bit easier to follow

Reply:
I'd like to kindly push back on the nested condition :( I agree that removing the anon function is cool, but my brain is really bad at following nested conditions or `else` statements (i.e. conditional conditions). This is why I always keep only one level of indentation, I found it to yield much much much easier to maintain code

Reply:
Still removed the anon function, thanks for the suggestion